compiler/ras/ILValidator.cpp

```diff
@@ -21,333 +21,132 @@
 
 #include "ras/ILValidator.hpp"
 
-#include "il/DataTypes.hpp"
-#include "il/Node.hpp"
-#include "il/Node_inlines.hpp"
-#include "il/ILProps.hpp"
-#include "il/ILOps.hpp"
-#include "il/Block.hpp"
-#include "il/Block_inlines.hpp"
-#include "infra/Assert.hpp"
-
-#include <stdarg.h>
-
-#if defined(DEBUG) || defined(PROD_WITH_ASSUMES)
-#define ABORT() TR::trap()
-#else
-#define ABORT() comp()->failCompilation<TR::CompilationException>("Validation error")
-#endif
-
-#define FAIL() if (!feGetEnv("TR_continueAfterValidationError")) ABORT()
+#include <algorithm>                               // for std::for_each
+
+#include "compile/Compilation.hpp"                 // for Compilation
+#include "infra/ILWalk.hpp"                        // for PostorderNodeOccurrenceIterator
+#include "ras/BlockValidationRules.hpp"            // for BlockValidationRules
+#include "ras/MethodValidationRules.hpp"           // for MethodValidationRules
+#include "ras/NodeValidationRules.hpp"             // for NodeValidationRules
+
+template <typename T, size_t N> static
+T* begin(T(&reqArray)[N])
+  {
+  return &reqArray[0];
+  }
+template <typename T, size_t N> static
+T* end(T(&reqArray)[N])
+   {
+   return &reqArray[0]+N;
+   }
 
 TR::ILValidator::ILValidator(TR::Compilation *comp)
    :_comp(comp)
-   ,_isValidSoFar(true)
-   ,_nodeStates(comp->trMemory())
-   ,_liveNodes(_nodeStates, comp->trMemory())
    {
+     // Initialize all the available rules in the constructor.
+     // It is during the `validate` call where we decide which subset
+     // of the available ones we should call.
+     // This also lets one specify the ordering in which the rules are validated.
+     TR::MethodValidationRule* temp_method_rules[] = { new  TR::SoundnessRule(_comp),
+                                                       new  TR::ValidateLivenessBoundaries(_comp) };
+
+     TR::BlockValidationRule* temp_block_rules[] = { new  TR::ValidateNodeRefCountWithinBlock(_comp) };
+
+     TR::NodeValidationRule* temp_node_rules[] = { new  TR::ValidateChildCount(_comp),
+                                                   new  TR::ValidateChildTypes(_comp),
+                                                   new  TR::Validate_ireturnReturnType(_comp) };
```

Comment:
I don't think we can have use of `new` without making it placement-`new` on to a TR memory type: See #2003.